torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -0,0 +1,455 @@
+#pragma once
+
+#include <torch/data/datasets/stateful.h>
+
+namespace torch {
+namespace data {
+namespace datasets {
+
+/// Interface for chunk reader, which performs data chunking and reading of
+/// entire chunks.
+///
+/// A chunk could be an entire file, such as an audio data file or an image,
+/// or part of a file in the case of a large text-file split based on seek
+/// positions.
+template <typename Chunk = std::vector<Example<>>>
+class ChunkDataReader {
+ public:
+  using ChunkType = Chunk;
+
+  /// Read an entire chunk.
+  virtual ChunkType read_chunk(size_t chunk_index) = 0;
+
+  /// Returns the number of chunks available in this reader.
+  virtual size_t chunk_count() = 0;
+
+  /// This will clear any internal state associate with this reader.
+  virtual void reset() = 0;
+};
+
+namespace detail {
+/// BatchDataBuffer manages a queue of UnwrappedBatchData. After a new chunk is
+/// loaded, BatchDataBuffer splits it into small batches and push them into the
+/// queue. When get_batch is called from data loader, it pops cached batches and
+/// return. If the cache is empty, it either waits to load more chunks or return
+/// null if all chunks are loaded.
+template <
+    typename UnwrappedBatch = std::vector<Example<>>,
+    typename ExampleSampler = samplers::RandomSampler>
+class BatchDataBuffer {
+ public:
+  using UnwrappedBatchType = UnwrappedBatch;
+  using BatchType = torch::optional<UnwrappedBatchType>;
+  using BatchRequestType = typename ExampleSampler::BatchRequestType;
+
+  BatchDataBuffer(
+      size_t num_chunks,
+      size_t batch_size,
+      ExampleSampler example_sampler,
```

Comment:
Since the lifetime of this object is bound to `ChunkDataset` that uses it, it would be better to take a sampler by reference. Those are mutable objects, so you should avoid copying them when you don't need to.

Reply:
I meant that you should remove the `std::move`, take a reference here, and keep a reference as a member of this class. You are unnecessarily assuming that the sampler can be copy-constructed, and are incurring extra costs for no reason.

Reply:
No, in that case the user might explicitly move it into the `ChunkDataset`, and you should move it into a member variable. So it's possible that there are no copies involved, except for this one.